server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -348,61 +349,71 @@ protected void doInternalExecute(Task task, BulkRequest bulkRequest, Executor ex
-        final ClusterState state = clusterService.state();
-        Map<String, Boolean> indicesToAutoCreate = indices.entrySet()
-            .stream()
-            .filter(entry -> indexNameExpressionResolver.hasIndexAbstraction(entry.getKey(), state) == false)
-            // We should only auto create if we are not requiring it to be an alias
-            .filter(entry -> entry.getValue().isRequireAlias == false)
-            .collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().isRequireDataStream));
-
-        // Step 3: Collect all the data streams that need to be rolled over before writing
-        Set<String> dataStreamsToBeRolledOver = featureService.clusterHasFeature(state, LazyRolloverAction.DATA_STREAM_LAZY_ROLLOVER)
-            ? indices.keySet().stream().filter(target -> {
-                DataStream dataStream = state.metadata().dataStreams().get(target);
-                return dataStream != null && dataStream.rolloverOnWrite();
-            }).collect(Collectors.toSet())
-            : Set.of();
-
-        // Step 4: create all the indices that are missing, if there are any missing. start the bulk after all the creates come back.
         createMissingIndicesAndIndexData(
             task,
             bulkRequest,
             executor,
             listener,
             indicesToAutoCreate,
             dataStreamsToBeRolledOver,
-            indicesThatCannotBeCreated,
+            failureStoresToBeRolledOver,
             startTime
         );
     }
 
-    /*
-     * This method is responsible for creating any missing indices, rolling over a data stream when needed and then
-     *  indexing the data in the BulkRequest
+    private void populateMissingTargets(
+        BulkRequest bulkRequest,
+        Map<String, Boolean> indicesToAutoCreate,
+        Set<String> dataStreamsToBeRolledOver,
+        Set<String> failureStoresToBeRolledOver
+    ) {
+        Map<String, Boolean> indexExistence = new HashMap<>();
+        ClusterState state = clusterService.state();
+        boolean lazyRolloverFeature = featureService.clusterHasFeature(state, LazyRolloverAction.DATA_STREAM_LAZY_ROLLOVER);
+        boolean lazyRolloverFailureStoreFeature = featureService.clusterHasFeature(state, LazyRolloverAction.FAILURE_STORE_LAZY_ROLLOVER);
+
+        for (DocWriteRequest<?> request : bulkRequest.requests) {
+            if (request.opType() == OpType.DELETE
+                && request.versionType() != VersionType.EXTERNAL
+                && request.versionType() != VersionType.EXTERNAL_GTE) {
+                continue;
+            }
+            Boolean indexExists = indexExistence.get(request.index());
+            if (indexExists == null) {
+                indexExists = indexNameExpressionResolver.hasIndexAbstraction(request.index(), state);
+                indexExistence.put(request.index(), indexExists);
+            }
+            if (indexExists == false && request.isRequireAlias() == false) {
```

Comment:
Is this correct? If we have any request in the bulk that requires alias, then in the previous logic we would not add the index to the auto creation. The new logic looks like as long as a single operation does not require an alias it will be added to the auto creation.

Reply:
Ah I think you're right, good catch! Any thoughts on a good solution to this? The best solution I can think of is what I did in [3b6a10a](https://github.com/elastic/elasticsearch/pull/108108/commits/3b6a10a18293eb680b3b9ea2b9374b200e8bd350), but it requires yet another `Set` allocation and makes the loop more complex...

Reply:
I think if I were doing this from scratch I would have added some fields to `ReducedRequestInfo` to track if an index request was being executed regularly and if it was a failure store operation. Then I'd use the merge function on the map collector beforehand to OR the flags together. When it would come time to generate the rollovers I'd have generated a rollover request for each set flag on the `ReducedRequestInfo` for an index.

Reply:
Yeah I thought about making changes to `ReducedRequestInfo` initially as well, but because that's an enum and not a record, I'd have to create all kinds of permutations which seemed confusing to me.

Reply:
Are you ok with the current implementation? Or do you want me to take another crack at it? I've spent some time weighing off different approaches already, so not sure how far I'd come.

Reply:
I think your latest revision is likely the best approach for this. Good stuff!